Sources/LanguageServerProtocol/Connection.swift

```diff
@@ -136,3 +137,29 @@ extension LocalConnection: Connection {
     return id
   }
 }
+
+extension Connection {
+  /// Send the given request to the connection and await its result.
+  ///
+  /// This method automatically sends a `CancelRequestNotification` to the
+  /// connection if the task it is executing in is being cancelled.
+  ///
+  /// - Warning: Because this message is `async`, it does not provide any ordering
+  ///   guarantees. If you need to gurantee that messages are sent in-order
+  ///   use the version with a completion handler.
+  public func send<R: RequestType>(_ request: R) async throws -> R.Response {
+    let requestIDWrapper = ThreadSafeBox<RequestID?>(initialValue: nil)
+    return try await withTaskCancellationHandler {
+      try await withCheckedThrowingContinuation { continuation in
+        let requestID = self.send(request) { result in
```

Comment:
Should this be checking for `Task.checkCancellation()` before sending the request? Asking because I noticed you're doing so in the sourcekitd use of `withTaskCancellationHandler`, and after looking at the docs for `withTaskCancellationHandler` it seems like if the task is already cancelled, it will call `onCancel` _before_ the operation.

Reply:
Oh, very good catch. We should definitely call `Task.checkCancellation()` here.